File: chartiq/src/main/java/com/chartiq/sdk/scriptmanager/ChartIQScriptManager.java
package com.chartiq.sdk.scriptmanager;

import com.chartiq.sdk.model.AggregationType;
import com.chartiq.sdk.model.DrawingTool;
import com.chartiq.sdk.model.OHLCParams;
import com.google.gson.Gson;

import org.json.JSONObject;

import java.util.Map;

import static com.chartiq.sdk.Util.buildArgumentStringFromArgs;

public class ChartIQScriptManager implements ScriptManager {

    private String mobileNameSpace = "CIQ.MobileBridge.";
    private static final String CHART_IQ_JS_OBJECT = "stxx.";

    @Override
    public String getDetermineOSScript() {
        String script = mobileNameSpace + "determineOs()";
        return script;
    }

    @Override
    public String getNativeQuoteFeedScript() {
        String script = mobileNameSpace + "nativeQuoteFeed(parameters, cb)";
        return script;
    }

    @Override
    public String getAddDrawingListenerScript() {
        String script = mobileNameSpace + "addDrawingListener()";
        return script;
    }

    @Override
    public String getAddLayoutListenerScript() {
        String script = mobileNameSpace + "addLayoutListener()";
        return script;
    }

    @Override
    public String getSetDataMethodScript(String symbol) {
        String functionName = "newChart";
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + symbol + ")";
        return script;
    }

    @Override
    public String getSetSymbolScript(String symbol) {
        String script = mobileNameSpace + "loadChart" + "(" + "\"" + symbol + "\"" + ")";
        return script;
    }

    @Override
    public String getDateFromTickScript() {
        String functionName = "dateFromTick";
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + 1 + ")";
        return script;
    }

    @Override
    public String getSetAccessibilityModeScript() {
        String script = mobileNameSpace + "accessibilityMode()";
        return script;
    }

    @Override
    public String getIsChartAvailableScript() {
        return null;
    }

    @Override
    public String getSetSymbolObjectScript(JSONObject symbolObject) {
        return null;
    }

    @Override
    public String getSetPeriodicityScript(int period, String interval, String timeUnit) {
        if (timeUnit == null) {
            timeUnit = "minute";
        }
        String args = buildArgumentStringFromArgs(period, interval, timeUnit);
        String script = mobileNameSpace + "setPeriodicity(" + period + ", \"" + interval + "\", \"" + timeUnit + "\");";
        return script;
    }

    @Override
    public String getGetChartNameScript() {
        return null;
    }

    @Override
    public String getPushDataScript(String symbol, OHLCParams[] data) {
        return null;
    }

    @Override
    public String getPushUpdateScript(OHLCParams[] data) {
        return null;
    }

    @Override
    public String getSetChartTypeScript(String chartType) {
        String functionName = "setChartType";
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + buildArgumentStringFromArgs(chartType) + ")";
        return script;
    }

    @Override
    public String getAddComparisonScript(String symbol, String hexColor, Boolean isComparison) {
        return null;
    }

    @Override
    public String getRemoveComparisonScript(String symbol) {
        return null;
    }

    @Override
    public String getClearChartScript() {
        return null;
    }

    @Override
    public String getSetChartScaleScript(String scale) {
        String functionName = "setChartScale";
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + scale + ")";
        return script;
    }

    @Override
    public String getAddStudyScript(String studyName, Map<String, Object> inputs, Map<String, Object> outputs, Map<String, Object> parameters) {
        String functionName = "addStudy";
        String script = functionName + "(" + buildArgumentStringFromArgs(studyName, inputs, outputs, parameters) + ")";
        return script;
    }

    @Override
    public String getRemoveStudyScript(String studyName) {
        String script = mobileNameSpace + "removeStudy" + "(" + "\"" + studyName + "\"" + ")";
        return script;
    }

    @Override
    public String getRemoveAllStudiesScript() {
        return null;
    }

    @Override
    public String getEnableCrosshairsScript() {
        String script = mobileNameSpace + "enableCrosshairs(true);";
        return script;
    }

    @Override
    public String getIsCrosshairsEnabledScript() {
        return null;
    }

    @Override
    public String getDisableCrosshairsScript() {
        String script = mobileNameSpace + "enableCrosshairs(false);";
        return script;
    }

    @Override
    public String getGetCrosshairsHUDDetailScript() {
        return null;
    }

    @Override
    public String getEnableDrawingScript(DrawingTool type) {
        String functionName = "changeVectorType";
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + buildArgumentStringFromArgs(type.getValue()) + ")";
        return script;
    }

    @Override
    public String getDisableDrawingScript() {
        String script = getEnableDrawingScript(DrawingTool.NONE);
        return script;
    }

    @Override
    public String getClearDrawingScript() {
        String functionName = "clearDrawings";
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + ")";
        return script;
    }

    @Override
    public String getSetDrawingParameterScript(String parameter, String value) {
        String script = mobileNameSpace + "setCurrentVectorParameters" + '(' + parameter + ", " + value + ')' + ';';
        return script;
    }

    @Override
    public String getSetStyleScript(String object, String parameter, String value) {
        return null;
    }

    @Override
    public String getSetThemeScript(String theme) {
        return null;
    }

    @Override
    public String getGetStudyListScript() {
        String script = mobileNameSpace + "getStudyList();";
        return script;
    }

    @Override
    public String getGetActiveStudiesScript() {
        String script = mobileNameSpace + "getActiveStudies();";
        return script;
    }

    @Override
    public String getSetAggregationTypeScript(AggregationType aggregationType) {
        String functionName = "setAggregationType";
        String value = aggregationType.getValue();
        String script = CHART_IQ_JS_OBJECT + functionName + "(" + value + ")";
        return script;
    }

    @Override
    public String getGetStudyInputParametersScript(String studyName) {
        String script = mobileNameSpace + "getStudyParameters(\"" + studyName + "\" , \"inputs\");";
        return script;
    }

    @Override
    public String getGetStudyOutputParametersScript(String studyName) {
        String script = mobileNameSpace + "getStudyParameters(\"" + studyName + "\" , \"outputs\");";
        return script;
    }

    @Override
    public String getGetStudyParametersScript(String studyName) {
        String script = mobileNameSpace + "getStudyParameters(\"" + studyName + "\" , \"parameters\");";
        return script;
    }

    @Override
    public String getSetStudyInputParameterScript(String studyName, String parameter, String value) {
        return null;
    }

    @Override
    public String getSetStudyOutputParameterScript(String studyName, String parameter, String value) {
        return null;
    }

    @Override
    public String getGetDrawingParametersScript(String drawingName) {
        return null;
    }

    @Override
    public String getChangeChartStyleScript(Object... args) {
        return null;
    }

    @Override
    public String getSetChartPropertyScript(String property, Object value) {
        return null;
    }

    @Override
    public String getGetChartPropertyScript(String property) {
        return null;
    }

    @Override
    public String getSetEnginePropertyScript(String property, Object value) {
        return null;
    }

    @Override
    public String getGetEnginePropertyScript(String property) {
        return null;
    }


    @Override
    public String getParseDataScript(OHLCParams[] data, String callbackId) {
        String json = new Gson().toJson(data);
        String script = mobileNameSpace + "parseData('" + json + "', \"" + callbackId + "\");";
        return script;
    }
}

File: chartiq/src/main/java/com/chartiq/sdk/scriptmanager/ChartIQScriptManager.kt
package com.chartiq.sdk.scriptmanager

import com.chartiq.sdk.model.ChartLayer
import com.chartiq.sdk.model.ChartTheme
import com.chartiq.sdk.model.OHLCParams
import com.chartiq.sdk.model.drawingtool.DrawingTool
import com.chartiq.sdk.model.study.StudyParameterModel
import com.google.gson.Gson

// TODO: 03.09.20 Add parameters safety check
internal class ChartIQScriptManager : ScriptManager {

    override fun getDetermineOSScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "determineOs()"

    override fun getNativeQuoteFeedScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "nativeQuoteFeed(parameters, cb)"

    override fun getAddDrawingListenerScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "addDrawingListener();"

    override fun getAddLayoutListenerScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "addLayoutListener()"

    override fun getAddMeasureListener(): String =
        MOBILE_BRIDGE_NAME_SPACE + "addMeasureListener();"

    override fun getSetSymbolScript(symbol: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "loadChart(\"$symbol\")"

    override fun getDateFromTickScript(): String =
        CHART_IQ_JS_OBJECT + "dateFromTick(${1})"

    override fun getLoadChartScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "loadChart();"

    override fun getSetAccessibilityModeScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "accessibilityMode();"

    override fun getIsChartAvailableScript(): String =
        "if (" + MOBILE_BRIDGE_NAME_SPACE + "isChartAvailable() == true) { \"true\" } else { \"false\" } "

    override fun getSetPeriodicityScript(period: Int, interval: String, timeUnit: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "setPeriodicity($period, $interval, \"$timeUnit\");"

    override fun getPushDataScript(symbol: String, data: List<OHLCParams>): String {
        // TODO: 03.09.20 Parse the array for the call
        val jsonString = ""
        return MOBILE_BRIDGE_NAME_SPACE + "loadChart(\"\", $jsonString); "
    }

    override fun getPushUpdateScript(data: List<OHLCParams>): String {
        // TODO: 03.09.20 Parse the array for the call
        return MOBILE_BRIDGE_NAME_SPACE + "parseData($data');"
    }

    override fun getSetChartTypeScript(chartType: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "setChartType(\"$chartType\");"

    override fun getSetAggregationTypeScript(aggregationType: String): String =
        CHART_IQ_JS_OBJECT + "setAggregationType(\"$aggregationType\");"

    override fun getChartTypeScript(): String = CHART_IQ_JS_OBJECT + "layout.chartType"

    override fun getAggregationTypeScript(): String = CHART_IQ_JS_OBJECT + "layout.aggregationType"

    override fun getResizeChartScript(): String =
        CHART_IQ_JS_OBJECT + "resizeChart();"

    override fun getClearChartScript(): String =
        CHART_IQ_JS_OBJECT + "destroy();"

    override fun getChartScaleScript(): String = CHART_IQ_JS_OBJECT + "layout.chartScale"

    override fun getSetChartScaleScript(scale: String): String =
        CHART_IQ_JS_OBJECT + "setChartScale(\"$scale\");"

    override fun getAddStudyScript(studyName: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "addStudy(\"$studyName\");"

    override fun getRemoveStudyScript(studyName: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "removeStudy(\"$studyName\");"

    override fun getRemoveAllStudiesScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "removeAllStudies();"

    override fun getEnableCrosshairScript(value: Boolean): String =
        MOBILE_BRIDGE_NAME_SPACE + "enableCrosshairs($value);"

    override fun getIsCrosshairsEnabledScript(): String =
        "if (${CHART_IQ_JS_OBJECT}layout.crosshair == true) { \"true\" } else { \"false\" } "

    override fun getGetCrosshairHUDDetailsScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "getHudDetails();"

    override fun getEnableDrawingScript(type: DrawingTool): String =
        "currentDrawing = \"${type.value}\";" + CHART_IQ_JS_OBJECT + "changeVectorType(currentDrawing);"

    override fun getDisableDrawingScript(): String = getEnableDrawingScript(DrawingTool.NONE)

    override fun getClearDrawingScript(): String = CHART_IQ_JS_OBJECT + "clearDrawings();"

    override fun getRestoreDefaultDrawingConfigScript(tool: DrawingTool, all: Boolean): String {
        var toolName = ""
        if (!all) {
            toolName = tool.value
        }
        return MOBILE_BRIDGE_NAME_SPACE + "restoreDefaultDrawingConfig(\"$toolName\", $all);"
    }

    override fun getSetDrawingParameterScript(parameter: String, value: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "setDrawingParameters(\"$parameter\", `$value`);"

    override fun getSetStyleScript(obj: String, parameter: String, value: String): String =
        CHART_IQ_JS_OBJECT + "setStyle(\"$obj\", \"$parameter\", \"$value\");"

    override fun getSetThemeScript(theme: ChartTheme): String =
        MOBILE_BRIDGE_NAME_SPACE + "setTheme(\"${theme.value}\");"

    override fun getGetStudyListScript(): String = "JSON.stringify(CIQ.Studies.studyLibrary);"

    override fun getGetActiveStudiesScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "getActiveStudies();"

    override fun getStudyInputParametersScript(studyName: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "getStudyParameters(\"$studyName\", \"inputs\")"

    override fun getStudyOutputParametersScript(studyName: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "getStudyParameters(\"$studyName\" , \"outputs\")"

    override fun getStudyParametersScript(studyName: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "getStudyParameters(\"$studyName\" , \"parameters\")"

    override fun getSetStudyParameterScript(
        studyName: String,
        parameter: StudyParameterModel
    ): String {
        return MOBILE_BRIDGE_NAME_SPACE + "setStudy(\"$studyName\", \"${parameter.fieldName.asSafeScriptParameter}\", \"${parameter.fieldSelectedValue.asSafeScriptParameter}\");"
    }

    override fun getSetStudyParametersScript(
        name: String,
        parameters: List<StudyParameterModel>
    ): String {
        val scriptList = parameters.map {
            getUpdateStudyParametersScript(it.fieldName, it.fieldSelectedValue)
        }
        return getStudyDescriptorScript(name) +
                "var helper = new CIQ.Studies.DialogHelper({\n" +
                "\tsd: selectedSd,\n" +
                "\tstx: stxx\n" +
                "});\n" +
                "var isFound = false;\n" +
                "var newInputParameters = {};\n" +
                "var newOutputParameters = {};\n" +
                "var newParameters = {};" +
                scriptList.joinToString(" ") +
                "helper.updateStudy({\n" +
                "\tinputs: newInputParameters,\n" +
                "\toutputs: newOutputParameters,\n" +
                "\tparameters: newParameters\n" +
                "});\n" +
                "console.log(JSON.stringify(newInputParameters));\n" +
                "console.log(JSON.stringify(newOutputParameters));\n" +
                "console.log(JSON.stringify(newParameters));"
    }

    private fun getStudyDescriptorScript(name: String): String {
        val safeStudyName = name.asSafeScriptParameter
        return "var s = stxx.layout.studies;\n" +
                "var selectedSd = {};\n" +
                "for (var n in s) {\n" +
                "\tvar sd = s[n];\n" +
                "\tif (sd.name === \"$safeStudyName\") {\n" +
                "\t\tselectedSd = sd;\n" +
                "\t}\n" +
                "}"

    }

    private fun getUpdateStudyParametersScript(key: String, value: String): String {
        val safeStudyParameter = key.asSafeScriptParameter
        val safeStudyValue = value.asSafeScriptParameter
        val script = "for (x in helper.inputs) {" +
                "   var input = helper.inputs[x]; " +
                "   if (input[\"name\"] === \"$safeStudyParameter\") { " +
                "       isFound = true; " +
                "       if (input[\"type\"] === \"text\" || input[\"type\"] === \"select\") { " +
                "           newInputParameters[\"$safeStudyParameter\"] = \"$safeStudyValue\"; " +
                "       } else if (input[\"type\"] === \"number\") { " +
                "           newInputParameters[\"$safeStudyParameter\"] = parseFloat(\"$safeStudyValue\"); " +
                "       } else if (input[\"type\"] === \"checkbox\") { " +
                "           newInputParameters[\"$safeStudyParameter\"] =" +
                " (\"$safeStudyValue\" == \"false\" || \"$safeStudyValue\" == \"0\" ? false : true); " +
                "       } " +
                "   } " +
                "} " +
                "if (isFound == false) { " +
                "   for (x in helper.outputs) { " +
                "       var output = helper.outputs[x]; " +
                "       if (output[\"name\"] === \"$safeStudyParameter\") { " +
                "           newOutputParameters[\"$safeStudyParameter\"] = \"$safeStudyValue\"; " +
                "       } " +
                "   } " +
                "} " +
                "if (isFound == false) { " +
                "   if(\"$safeStudyParameter\".includes(\"Color\")) { " +
                "       newParameters[\"$safeStudyParameter\"] = \"$safeStudyValue\"; " +
                "   } else if(\"$safeStudyParameter\".includes(\"Enabled\")) { " +
                "       newParameters[\"$safeStudyParameter\"] =" +
                " (\"$safeStudyValue\" == \"false\" || \"$safeStudyValue\" == \"0\" ? false : true); " +
                "   } else { " +
                "       newParameters[\"$safeStudyParameter\"] = parseFloat(\"$safeStudyValue\"); " +
                "   } " +
                "} " + "isFound = false; "
        return script
    }

    override fun getGetDrawingParametersScript(drawingName: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "getDrawingParameters(\"$drawingName\");"

    override fun getSetChartStyleScript(obj: String, attribute: String, value: String): String =
        CHART_IQ_JS_OBJECT + "setStyle(\"$obj\", \"$attribute\", \"$value\")"

    override fun getSetChartPropertyScript(property: String, value: Any): String {
        TODO("Not yet implemented")
    }

    override fun getGetChartPropertyScript(property: String): String {
        TODO("Not yet implemented")
    }

    override fun getSetEnginePropertyScript(property: String, value: Any): String {
        TODO("Not yet implemented")
    }

    override fun getGetEnginePropertyScript(property: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "getEngineProperty(\"$property\");"

    override fun getParseDataScript(data: List<OHLCParams>, callbackId: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "parseData('${Gson().toJson(data)}', \"$callbackId\")"

    override fun getInvertYAxisScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "getLayoutProperty(\"flipped\");"

    override fun getSetInvertYAxisScript(inverted: Boolean): String = CHART_IQ_JS_OBJECT + "flipChart($inverted);"

    override fun getIsExtendedHoursScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "getLayoutProperty(\"extended\");"

    override fun getSetExtendedHoursScript(extended: Boolean): String =
        CHART_IQ_JS_OBJECT + "extendedHours.set($extended);"

    override fun getUndoDrawingScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "undo();"

    override fun getRedoDrawingScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "redo();"

    override fun getDeleteDrawingScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "deleteDrawing();"

    override fun getCloneDrawingScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "cloneDrawing();"

    override fun getLayerManagementScript(layer: ChartLayer): String =
        MOBILE_BRIDGE_NAME_SPACE + "layerDrawing(\"${layer.value}\");"

    override fun getGetTranslationsScript(languageCode: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "getTranslations(\"${languageCode.asSafeScriptParameter}\");"

    override fun getSetLanguageScript(languageCode: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "setLanguage(\"${languageCode.asSafeScriptParameter}\");"

    override fun getGetActiveSeriesScript(): String =
        MOBILE_BRIDGE_NAME_SPACE + "getAllSeries()"

    override fun getAddSeriesScript(symbol: String, color: String, isComparison: Boolean): String =
        CHART_IQ_JS_OBJECT + "addSeries(\"$symbol\", {display:\"$symbol\", color: \"$color\", isComparison:\"$isComparison\"});"

    override fun getRemoveSeriesScript(symbol: String): String =
        CHART_IQ_JS_OBJECT + "removeSeries(\"$symbol\");"

    override fun getSetSeriesParameterScript(symbol: String, field: String, value: String): String =
        MOBILE_BRIDGE_NAME_SPACE + "modifySeries(\"$symbol\", \"$field\", \"$value\");"

    private val String.asSafeScriptParameter: String
        get() {
            //todo check how it works
            return this.replace("(?i)(<.?\\s+)on.?(>.*?)", "$1$2")
        }

    companion object {
        private const val MOBILE_BRIDGE_NAME_SPACE = "CIQ.MobileBridge."
        private const val CHART_IQ_JS_OBJECT = "stxx."
    }
}
